src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php

```diff
@@ -4,142 +4,183 @@
-        $this->scheduledDelayMillis = $scheduledDelayMillis
-            ?: $this->getIntFromEnvironment(Env::OTEL_BSP_SCHEDULE_DELAY, self::DEFAULT_SCHEDULE_DELAY);
-        $this->exporterTimeoutMillis = $exporterTimeoutMillis
-            ?: $this->getIntFromEnvironment(Env::OTEL_BSP_EXPORT_TIMEOUT, self::DEFAULT_EXPORT_TIMEOUT);
-        $this->maxExportBatchSize = $maxExportBatchSize
-            ?: $this->getIntFromEnvironment(Env::OTEL_BSP_MAX_EXPORT_BATCH_SIZE, self::DEFAULT_MAX_EXPORT_BATCH_SIZE);
-        if ($this->maxExportBatchSize > $this->maxQueueSize) {
-            throw new InvalidArgumentException(
-                sprintf('maxExportBatchSize should be smaller or equal to %s', $this->maxQueueSize)
-            );
+        if ($maxQueueSize <= 0) {
+            throw new InvalidArgumentException(sprintf('Maximum queue size (%d) must be greater than zero', $maxQueueSize));
+        }
+        if ($scheduledDelayMillis <= 0) {
+            throw new InvalidArgumentException(sprintf('Scheduled delay (%d) must be greater than zero', $scheduledDelayMillis));
+        }
+        if ($exportTimeoutMillis <= 0) {
+            throw new InvalidArgumentException(sprintf('Export timeout (%d) must be greater than zero', $exportTimeoutMillis));
+        }
+        if ($maxExportBatchSize <= 0) {
+            throw new InvalidArgumentException(sprintf('Maximum export batch size (%d) must be greater than zero', $maxExportBatchSize));
         }
+        if ($maxExportBatchSize > $maxQueueSize) {
+            throw new InvalidArgumentException(sprintf('Maximum export batch size (%d) must be less than or equal to maximum queue size (%d)', $maxExportBatchSize, $maxQueueSize));
+        }
+
+        $this->exporter = $exporter;
+        $this->clock = $clock;
+        $this->maxQueueSize = intdiv($maxQueueSize, $maxExportBatchSize);
+        $this->maxBatchSize = $maxQueueSize % $maxExportBatchSize;
+        $this->scheduledDelayNanos = $scheduledDelayMillis * 1_000_000;
+        $this->maxExportBatchSize = $maxExportBatchSize;
+        $this->autoFlush = $autoFlush;
+
+        $this->queue = new SplQueue();
+        $this->flush = new SplQueue();
     }
 
-    /**
-     * @inheritDoc
-     */
     public function onStart(ReadWriteSpanInterface $span, Context $parentContext): void
     {
     }
 
-    /**
-     * @inheritDoc
-     */
     public function onEnd(ReadableSpanInterface $span): void
     {
-        if (null === $this->exporter) {
+        if ($this->closed) {
             return;
         }
-
-        if (!$this->running) {
+        if (!$span->getContext()->isSampled()) {
             return;
         }
 
-        if ($span->getContext()->isSampled() && !$this->queueReachedLimit()) {
-            $this->queue[] = $span->toSpanData();
+        if (count($this->queue) === $this->maxQueueSize && count($this->batch) === $this->maxBatchSize) {
```

Comment:
should this be `>=` ?

Reply:
This defines the upper limit of the queue and cannot be exceeded.

Side note: Currently it allows `maxQueueSize` + up to `maxExportBatchSize` (size of in-flight batch) not exported spans, will change it to consider in-flight batches for consistency. We need this change anyways as soon as we add concurrent exports in a follow up PR.